### PR TITLE
Pause the video when the page is not visible.

### PR DESCRIPTION
**Pull request by shazar**
I've added a feature to pause the video when the page is not visible. If that looks good I'd appreciate if you could merge this.

Thanks.

### CODE REVIEW

File: src/jquery.vide.js [hunk 1]
```diff
@@ -32,9 +32,13 @@
     autoplay: true,
     position: '50% 50%',
     posterType: 'detect',
-    resizing: true
+    resizing: true,
+    pauseWhenHidden: true
```
Comment (VodkaBears) [inline src/jquery.vide.js]: My opinion is that this feature should be default or it should not be at all.

File: src/jquery.vide.js [hunk 2]
```diff
@@ -161,6 +165,51 @@
   }
 
   /**
+   * Find if there's support for page visibility.
+   * @private
+   * @returns null on no support, array of two elements if supported such that:
+   *   [hidden property name, visibility change event name]
+   */
+  function detectVisibilitySupport() {
+    var prefixes = ['webkit', 'moz', 'ms', 'o'];
```
Comment (VodkaBears) [inline src/jquery.vide.js]: "o" prefix is unnecessary

File: src/jquery.vide.js [hunk 3]
```diff
@@ -161,6 +165,51 @@
...
+  /**
+   * Play or pause video depending on the visibility state
+   * @private
+   * @param {Video} video
+   */
+  function handleVisibilityChange(video) {
+    if (document[hiddenPropertyName]) {
+      video.pause();
+    } else {
+      video.play();
```
Comment (VodkaBears) [inline src/jquery.vide.js]: Wrong behavior if the video was paused.

File: src/jquery.vide.js [hunk 4]
```diff
@@ -313,6 +362,17 @@
       // Resize a video, when it's loaded
       .one('canplaythrough.' + PLUGIN_NAME, function() {
         vide.resize();
+
+        if (vide.settings.pauseWhenHidden) {
```
Comment (VodkaBears) [inline src/jquery.vide.js]: You do the same actions for each instance.

File: src/jquery.vide.js [hunk 5]
```diff
@@ -161,6 +165,51 @@
   }
 
   /**
+   * Find if there's support for page visibility.
+   * @private
+   * @returns null on no support, array of two elements if supported such that:
```
Comment (VodkaBears) [inline src/jquery.vide.js]: Bad JSDoc

Comment (VodkaBears) [inline src/jquery.vide.js]: @returns {Array|null}

File: src/jquery.vide.js [hunk 6]
```diff
@@ -161,6 +165,51 @@
...
+   *   [hidden property name, visibility change event name]
+   */
+  function detectVisibilitySupport() {
```
Comment (VodkaBears) [inline src/jquery.vide.js]: I don't like this function^^

File: src/jquery.vide.js [hunk 7]
```diff
@@ -313,6 +362,17 @@
       // Resize a video, when it's loaded
       .one('canplaythrough.' + PLUGIN_NAME, function() {
         vide.resize();
+
+        if (vide.settings.pauseWhenHidden) {
+          var visibilitySupport = detectVisibilitySupport();
+
+          if (visibilitySupport) {
+            hiddenPropertyName = visibilitySupport[0];
+            $(document).on(visibilitySupport[1], function() {
```
Comment (VodkaBears) [inline src/jquery.vide.js]: `$(document).on('webkitvisibilitychange mozvisibilitychange msvisibilitychange visibilitychange'`

File: src/jquery.vide.js [hunk 8]
```diff
@@ -161,6 +165,51 @@
...
+      return null;
+    }
+
+    return [propName, eventName];
+  }
+
+  /**
+   * Play or pause video depending on the visibility state
+   * @private
+   * @param {Video} video
```
Comment (VodkaBears) [inline src/jquery.vide.js]: @param {HTMLVideoElement}

File: src/jquery.vide.js (hunk 7 above)
Comment (VodkaBears) [inline src/jquery.vide.js]: And as i wrote above, the listener must be shared across the instances.

Comment (VodkaBears) [inline src/jquery.vide.js]: Prefixes for this event should be a constant value.

File: src/jquery.vide.js [hunk 9]
```diff
@@ -313,6 +362,17 @@
...
       .one('canplaythrough.' + PLUGIN_NAME, function() {
         vide.resize();
+
+        if (vide.settings.pauseWhenHidden) {
+          var visibilitySupport = detectVisibilitySupport();
+
+          if (visibilitySupport) {
+            hiddenPropertyName = visibilitySupport[0];
+            $(document).on(visibilitySupport[1], function() {
+              handleVisibilityChange(vide.$video[0]);
```
Comment (VodkaBears) [inline src/jquery.vide.js]: The logic of the function is not complecated, seems like `handleVisibilityChange` is unnecessary.

**Comment by shazar** [comment]
I've gone through your comments and you're right, it is stupid that the code is executed for each instance.  I've rewritten it based on your feedback. If you're interested in adding this functionality I can send a follow-up patch.
Thanks.

**Comment by VodkaBears** [comment]
Let's freeze it.

**Pull request closed without merging**